packages/contracts/contracts/MarginlyPoolWrapper.sol

```diff
@@ -43,35 +47,87 @@ contract MarginlyPoolWrapper is IMarginlyPoolWrapper {
-    TransferHelper.safeTransferFrom(baseToken, msg.sender, address(this), depositBaseAmount);
-    TransferHelper.safeApprove(baseToken, marginlyPoolAddress, depositBaseAmount);
-    marginlyPool.depositBase(depositBaseAmount);
+    depositBaseInternal(marginlyPool, amount);
+    marginlyPool.transferPosition(msg.sender);
+  }
+
+  /// @inheritdoc IMarginlyPoolWrapper
+  function long(address marginlyPoolAddress, uint256 depositBaseAmount, uint256 longBaseAmount) external payable lock {
+    require(whitelistedMarginlyPools[marginlyPoolAddress], 'NW'); // not whitelisted
+
+    IMarginlyPool marginlyPool = IMarginlyPool(marginlyPoolAddress);
+    depositBaseInternal(marginlyPool, depositBaseAmount);
     marginlyPool.long(longBaseAmount);
     marginlyPool.transferPosition(msg.sender);
   }
 
   /// @inheritdoc IMarginlyPoolWrapper
-  function short(address marginlyPoolAddress, uint256 depositQuoteAmount, uint256 shortBaseAmount) external lock {
+  function depositQuote(address marginlyPoolAddress, uint256 amount) external payable lock {
     require(whitelistedMarginlyPools[marginlyPoolAddress], 'NW'); // not whitelisted
+
     IMarginlyPool marginlyPool = IMarginlyPool(marginlyPoolAddress);
-    address quoteToken = marginlyPool.quoteToken();
-    TransferHelper.safeTransferFrom(quoteToken, msg.sender, address(this), depositQuoteAmount);
-    TransferHelper.safeApprove(quoteToken, marginlyPoolAddress, depositQuoteAmount);
-    marginlyPool.depositQuote(depositQuoteAmount);
+    depositQuoteInternal(marginlyPool, amount);
+    marginlyPool.transferPosition(msg.sender);
+  }
+
+  /// @inheritdoc IMarginlyPoolWrapper
+  function short(
+    address marginlyPoolAddress,
+    uint256 depositQuoteAmount,
+    uint256 shortBaseAmount
+  ) external payable lock {
+    require(whitelistedMarginlyPools[marginlyPoolAddress], 'NW'); // not whitelisted
+
+    IMarginlyPool marginlyPool = IMarginlyPool(marginlyPoolAddress);
+    depositQuoteInternal(marginlyPool, depositQuoteAmount);
     marginlyPool.short(shortBaseAmount);
     marginlyPool.transferPosition(msg.sender);
   }
 
+  /// @dev Internal version without lock modifier
+  function depositBaseInternal(IMarginlyPool marginlyPool, uint256 amount) internal {
+    address baseToken = marginlyPool.baseToken();
+    wrapEthOrTransferErc20(baseToken, amount);
+    TransferHelper.safeApprove(baseToken, address(marginlyPool), amount);
+    marginlyPool.depositBase(amount);
+  }
+
+  /// @dev Internal version without lock modifier
+  function depositQuoteInternal(IMarginlyPool marginlyPool, uint256 amount) internal {
+    address quoteToken = marginlyPool.quoteToken();
+    wrapEthOrTransferErc20(quoteToken, amount);
+    TransferHelper.safeApprove(quoteToken, address(marginlyPool), amount);
+    marginlyPool.depositQuote(amount);
+  }
+
+  function wrapEthOrTransferErc20(address token, uint256 amount) internal {
+    if (token == WETH9 && address(this).balance >= amount) {
```

Comment:
```
require(token != address(0));
```

Reply:
It is an internal method, token addresses come from the pool and couldn't be address(0)